File: web/client/src/app/model/account-transfer-export/exporter-base.ts
import { PaymentAgency } from 'src/app/model/payment-agency.model';
import { Company } from 'src/app/model/company.model';
import { StringUtil } from "src/app/common/util/string-util";
import { AccountTransferDetail } from 'src/app/model/account-transfer-detail.model';
import { Observable } from 'rxjs';
import { FixedField } from 'src/app/model/account-transfer-export/fixed-field';
import { LINE_FEED_CODE } from 'src/app/common/const/report-setting-kbn.const';

export class ExporterBase{

    public DivisionHeader:number = 1;
    public DivisionData:number = 2;
    public DivisionTrailer:number = 8;
    public DivisionEnd:number = 9;

    public PaymentAgency:PaymentAgency;
    public Company:Company;
    public DueAt:string;
    public DueAt2nd:string;

    public TotalCount:number=0;
    public TotalAmount:number=0;

    public Export(source:AccountTransferDetail[]): string{

        let exportData:Array<any> = [];
        exportData.push(this.GetHeaderRecord());

        source.forEach(detail =>{
            this.TotalCount++;
            this.TotalAmount += detail.billingAmount;
            exportData.push(this.GetDataRecord(detail));
        });

        exportData.push(this.GetTrailerRecord());
        exportData.push(this.GetEndRecord());

        return exportData.join("");
    }

    public GetHeaderRecord():string{
        return "";
    }

    public GetDataRecord(detail:AccountTransferDetail):string{
        return "";
    }

    public GetTrailerRecord():string{
        return "";
    }

    public GetEndRecord():string{
        return "";
    }

    protected JoinData(dataItems: Array<any>):string{
        return dataItems.join(",") + LINE_FEED_CODE;
    }

    protected JoinFixedData(dataItems: Array<any>):string{
        return dataItems.join("") + LINE_FEED_CODE;
    }

    protected GetStrField(length:number, value:any, padding:string = ' ', rightPadding:boolean = true):string{
        let fixedField:FixedField = new FixedField(length, value, padding, rightPadding);
        return fixedField.ToString();
    }

    protected GetNmbField(length:number, value:any):string{
        let fixedField: FixedField = new FixedField(length, value, '0', false);
        return fixedField.ToString();
    }

    protected Dummy(length:number = 0):string{
        return length == 0 ? "" : String(' ').repeat(length);
    }

    protected GetMMdd(dueDate:string):string{
        let tDate = new Date(dueDate);
        let month:string = ""+ (tDate.getMonth()+1);
        let day:string = ""+ tDate.getDate();
        return StringUtil.setPaddingFrontZero(month, 2, true) + StringUtil.setPaddingFrontZero(day, 2, true);
    }


}

File: src/Rac.VOne.AccountTransfer/Export/ExporterBase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Rac.VOne.Web.Models;

namespace Rac.VOne.AccountTransfer.Export
{
    /// <summary>
    /// 口座振替依頼データ作成 base class
    /// </summary>
    public abstract class ExporterBase : IAccountTransferExporter
    {
        /// <summary>データ区分 1 : ヘッダーレコード</summary>
        protected const int DivisionHeader = 1;
        /// <summary>データ区分 2 : データレコード</summary>
        protected const int DivisionData = 2;
        /// <summary>データ区分 8 : トレーラーレコード</summary>
        protected const int DivisionTrailer = 8;
        /// <summary>データ区分 9 : エンドレコード</summary>
        protected const int DivisionEnd = 9;

        public Action<Exception> LogError { get; set; }
        public PaymentAgency PaymentAgency { get; set; }
        public Company Company { get; set; }
        public int TotalCount { get; set; }
        public decimal TotalAmount { get; set; }
        public Encoding Encoding { get; set; } = Encoding.GetEncoding(932);
        public DateTime DueAt { get; set; }
        public virtual void Export(string path, IEnumerable<AccountTransferDetail> source)
        {
            var builder = new StringBuilder();
            builder.Append(GetHeaderRecord());
            foreach (var detail in source)
            {
                TotalCount++;
                TotalAmount += detail.BillingAmount;
                builder.Append(GetDataRecord(detail));
            }
            builder.Append(GetTrailerRecord());
            builder.Append(GetEndRecord());
            try
            {
                System.IO.File.WriteAllText(path, builder.ToString(), Encoding);
            }
            catch (Exception ex)
            {
                LogError?.Invoke(ex);
            }
        }
        public virtual string GetHeaderRecord() => string.Empty;
        public virtual string GetDataRecord(AccountTransferDetail detail) => string.Empty;
        public virtual string GetTrailerRecord() => string.Empty;
        public virtual string GetEndRecord() => string.Empty;

        protected string Join(params object[] args)
            => string.Join(",", args) + Environment.NewLine;

        protected string Join(params FixedField[] args)
            => string.Concat(args.Select(x => x.ToString())) + Environment.NewLine;
        protected FixedField GetStrField(int length, object value, char padding = ' ', bool rightPadding = true)
            => new FixedField(length, value, padding, rightPadding);
        protected FixedField GetNmbField(int length, object value)
            => new FixedField(length, value, '0', false);
        /// <summary>
        /// 空文字 or 指定した文字数分 半角スペース
        /// </summary>
        /// <param name="length"></param>
        /// <returns></returns>
        protected string Dummy(int length = 0)
            => length == 0 ? string.Empty : new string(' ', length);
    }
}
